### PR TITLE
Closes #4988 Update delay js script & HTML markup

### LINKED ISSUE #4988: Update Delay JS script to latest version
**Is your feature request related to a problem? Please describe.**
We have 3 fixes left on the Delay JS script:
- https://github.com/wp-media/delay-javascript-loading/pull/18
- https://github.com/wp-media/delay-javascript-loading/pull/21
- https://github.com/wp-media/delay-javascript-loading/pull/24

**Describe the solution you'd like**
Merge all PR to create a new version of the script.

**Additional context**
PR 24 needs some change inside WP Rocket too.

### PR DESCRIPTION
**Pull request by jeawhanlee**
## Description

This PR updates the delay js script and set src to data-rocket-src for compatibility with update script.

Fixes #4988 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Is the solution different from the one proposed during the grooming?

This is solution is no way different from the proposed one.

## How Has This Been Tested?

Manually

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings

### CODE REVIEW

**Review by remyperona — APPROVED** [review]

**Comment by piotrbak** [comment]
As per https://github.com/wp-media/delay-javascript-loading/pull/32#issuecomment-1187002433, we need to make sure that only types mentioned Gael are rewritten to DelayJS markup.

**[EDIT]** I can see it being done, great!

**Comment by Mai-Saad** [comment]
@jeawhanlee @piotrbak 
- We have the following **regression**:
If we used this markup `<script type='text/javascript' src='https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/scripts/print-helloworld.js'></script>`, it is not delayed. which resulted in a js error while using [Avada theme](https://mega.wp-rocket.me/avada/) (this was working before)
Screenshot from 2022-08-26 15-57-53
Screenshot from 2022-08-26 15-57-30


**Notes** 

1. if `type='text/javascript'` came after the script, it will be delayed i.e `<script  src='https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/scripts/print-helloworld.js' type='text/javascript'></script>`
2. if we used before script `Type='text/javascript'` or `TYPE='text/javascript'`, will be normally delayed

**Question**
we are dealing with this markup
```
<script type="" src='https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/scripts/print-helloworld-v2.js'></script>
<script type src='https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/scripts/print-helloworld-v2.js'></script>
```
as
```
<script type="" src='https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/scripts/print-helloworld-v2.js'></script>
<script type="rocketlazyloadscript" type data-rocket-src='https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/scripts/print-helloworld-v2.js'></script>
```
shouldn't we delay 1st script and not duplicate the type in the 2nd script?

**Comment by piotrbak** [comment]
@Mai-Saad Regarding the **Question**, it seems to be an edge case. 

Could you confirm that:
1. This markup (not cached) is valid HTML markup
2. The first script `type=""` is executed normally

**Comment by Mai-Saad** [comment]
As per discussion with @piotrbak and further investigations:
- The first script `type=""` is executed normally in the browser so probably we need to handle
- The 2nd script `type`, we shouldn't add another type as having 2 types will result in an error using validator [HTML](https://validator.w3.org/nu/?doc=https%3A%2F%2Fnew.rocketlabsqa.ovh%2Ftest-browser%2F%3Fnowprocket)

**Comment by jeawhanlee** [comment]
> As per discussion with @piotrbak and further investigations:
> 
> * The first script `type=""` is executed normally in the browser so probably we need to handle
> * The 2nd script `type`, we shouldn't add another type as having 2 types will result in an error using validator [HTML](https://validator.w3.org/nu/?doc=https%3A%2F%2Fnew.rocketlabsqa.ovh%2Ftest-browser%2F%3Fnowprocket)

@Mai-Saad I was not able to reproduce the duplicate type using the markup:
`<script type src='https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/scripts/print-helloworld-v2.js'></script>`

File: inc/Engine/Preload/Admin/Settings.php [hunk 1]
```diff
@@ -128,10 +128,11 @@ private function is_valid_as_tables() {
 		}
 
 		global $wpdb;
-
+		$exp = "'^" . $wpdb->prefix . "actionscheduler_+(logs|actions|groups|claims)$'";
```
Comment (mostafa-hisham) [inline inc/Engine/Preload/Admin/Settings.php]: there is a conflict here with develop
`is_valid_as_tables` function are not in develop anymore

Comment (jeawhanlee) [reply inc/Engine/Preload/Admin/Settings.php]: Thanks @mostafa-hisham

Comment (jeawhanlee) [reply inc/Engine/Preload/Admin/Settings.php]: Thanks @mostafa-hisham

**Review by mostafa-hisham — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@jeawhanlee @piotrbak Seems we have regression on Firefox, so far while delay JS is enabled:
- Using the bridge theme, for some reason, the home page slider is not loaded after user interaction http://mega.wp-rocket.me/bridge
- Using Divi, the page keeps loading forever after user interaction http://mega.wp-rocket.me/divi/
can you please check?


https://user-images.githubusercontent.com/76941962/191034777-e50acbea-0369-400d-bc13-5cb8405c3564.mp4

**Review by mostafa-hisham — APPROVED** [review]

**Review by mostafa-hisham — APPROVED** [review]

File: inc/Engine/Optimization/DelayJS/HTML.php [hunk 2]
```diff
@@ -169,21 +194,46 @@ public function replace_scripts( $matches ): string {
...
+				$matches['attr'] = preg_replace( $type_exp, '', $matches['attr'], 1 );
+				$matches[0]      = preg_replace( '/\s\s+/', ' ', preg_replace( $type_exp, '', $matches[0], 1 ) );
+				$delay_js        = $matches[0];
+			}
+
+			// Check and remove dissociate type attribute.
+			if ( ! preg_match( $type_exp, $matches['attr'] ) && (bool) preg_match_all( '/type([^\w\W]|\s)|type$/i', $matches['attr'] ) ) {
+				$altered_attr = preg_replace( '/type([^\w\W]|\s)|type$/i', '', $matches['attr'] );
+
+				$delay_js = preg_replace( '/\s\s+/', ' ', preg_replace( '#' . preg_quote( $matches['attr'], '#' ) . '#i', $altered_attr, $matches[0], 1 ) );
```
Comment (remyperona) [inline inc/Engine/Optimization/DelayJS/HTML.php]: that's a lot of regex usage, is there any way to simplify this part? @wp-media/php any ideas?

Comment (piotrbak) [reply inc/Engine/Optimization/DelayJS/HTML.php]: @wp-media/php What do you think about @Tabrisrp comment here?

Comment (piotrbak) [reply inc/Engine/Optimization/DelayJS/HTML.php]: @wp-media/php Could you check that comment? What do you think?

Comment (remyperona) [reply inc/Engine/Optimization/DelayJS/HTML.php]: also, what cases are we trying to handle with those?

Comment (jeawhanlee) [reply inc/Engine/Optimization/DelayJS/HTML.php]: This [here](https://github.com/wp-media/wp-rocket/pull/5256#issuecomment-1228537288) and [here](https://github.com/wp-media/wp-rocket/pull/5256#issuecomment-1231804760)

**Review by remyperona — APPROVED** [review]

**Comment by remyperona** [comment]
PR needs to be updated with develop branch, and conflict fixed

**Comment by engahmeds3ed** [comment]
@Tabrisrp all good now, thanks for checking.

File: inc/Engine/Optimization/DelayJS/HTML.php [hunk 3]
```diff
@@ -52,6 +52,7 @@ class HTML {
 		'text/javascript1.4',
 		'text/javascript1.5',
 		'text/jscript',
+		'application/jscript',
```
Comment (remyperona) [inline inc/Engine/Optimization/DelayJS/HTML.php]: this is not a valid type

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
As per the exploratory test, we have regression on FF mobile view => page keeps loading and has many warnings while a separate mobile cache is enabled/disabled and Delay JS has no exclusions. @engahmeds3ed @jeawhanlee can you please check :pray: 

https://user-images.githubusercontent.com/76941962/205810699-a8b1450d-7da8-448c-943e-b7c2720bdc01.mp4

Note: Same is happening on mega/divi (working fine on the trunk)

**Comment by piotrbak** [comment]
For Avada example, excluding the following scripts makes slider appear right away. The root cause is most likely inside them:
```
/wp-content/themes/Avada/includes/lib/assets/min/js/library/jquery.flexslider.js
/avada/wp-includes/js/jquery/jquery.min.js
/wp-content/plugins/fusion-core/js/min/avada-fusion-slider.js
/wp-content/themes/Avada/includes/lib/assets/min/js/library/modernizr.js
/wp-content/themes/Avada/includes/lib/assets/min/js/general/fusion.js
/wp-content/themes/Avada/includes/lib/assets/min/js/library/cssua.js
/wp-content/themes/Avada/includes/lib/assets/min/js/library/bootstrap.transition.js
/wp-content/themes/Avada/includes/lib/assets/min/js/library/bootstrap.tooltip.js
```

**Comment by gmetais** [comment]
> As per the exploratory test, we have regression on FF mobile view

@Mai-Saad @piotrbak Actually, this is not really a bug.

We used to have an issue on Firefox, which led me to implement a Firefox browser detection in the code by checking the user agent. But when you switch Firefox to mobile view, it overwrites the user agent and replaces it with an Android or iOS user agent.

It should not reproduce on a real Firefox mobile browser. If you still want to test, Firefox allows you to edit the user agent. Adding "Firefox/107.0" to the end of the user agent field fixes the issue.

Capture d’écran 2022-12-12 à 09 27 48 min


In case you still consider this issue as blocking (for example if some website creators use the Firefox mobile view to debug their website), we could try to find a different way to identify Firefox. For example, on Firefox, navigator.vendor === "" whereas other browsers have non-empty values. Kind of a hack and I prefer not to add it.

**Review by Mai-Saad — APPROVED** [review]
Working as expected.
[testrail-report-432.pdf](https://github.com/wp-media/wp-rocket/files/10231469/testrail-report-432.pdf)

**Pull request merged**